aws/rust-runtime/aws-runtime/src/request_info.rs

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_smithy_runtime::client::orchestrator::interceptors::{RequestAttempts, ServiceClockSkew};
+use aws_smithy_runtime_api::client::interceptors::{BoxError, Interceptor, InterceptorContext};
+use aws_smithy_runtime_api::client::orchestrator::{HttpRequest, HttpResponse};
+use aws_smithy_runtime_api::config_bag::ConfigBag;
+use aws_smithy_types::date_time::Format;
+use aws_smithy_types::retry::RetryConfig;
+use aws_smithy_types::timeout::TimeoutConfig;
+use aws_smithy_types::DateTime;
+use http::{HeaderName, HeaderValue};
+use std::borrow::Cow;
+use std::time::{Duration, SystemTime};
+
+#[allow(clippy::declare_interior_mutable_const)] // we will never mutate this
+const AMZ_SDK_REQUEST: HeaderName = HeaderName::from_static("amz-sdk-request");
+
+/// Generates and attaches a request header that communicates request-related metadata.
+/// Examples include:
+///
+/// - When the client will time out this request.
+/// - How many times the request has been retried.
+/// - The maximum number of retries that the client will attempt.
+#[non_exhaustive]
+#[derive(Debug, Default)]
+pub struct RequestInfoInterceptor {}
+
+impl RequestInfoInterceptor {
+    /// Creates a new `RequestInfoInterceptor`
+    pub fn new() -> Self {
+        RequestInfoInterceptor {}
+    }
+}
+
+impl RequestInfoInterceptor {
+    fn build_attempts_pair(
+        &self,
+        cfg: &ConfigBag,
+    ) -> Option<(Cow<'static, str>, Cow<'static, str>)> {
+        let request_attempts = cfg
+            .get::<RequestAttempts>()
+            .map(|r_a| r_a.attempts())
+            .unwrap_or(1);
+        let request_attempts = request_attempts.to_string();
+        Some((Cow::Borrowed("attempt"), Cow::Owned(request_attempts)))
+    }
+
+    fn build_max_attempts_pair(
+        &self,
+        cfg: &ConfigBag,
+    ) -> Option<(Cow<'static, str>, Cow<'static, str>)> {
+        // TODO(orchestrator_retries) What config will we actually store in the bag? Will it be a whole config or just the max_attempts part?
+        if let Some(retry_config) = cfg.get::<RetryConfig>() {
+            let max_attempts = retry_config.max_attempts().to_string();
+            Some((Cow::Borrowed("max"), Cow::Owned(max_attempts)))
+        } else {
+            None
+        }
+    }
+
+    fn build_ttl_pair(&self, cfg: &ConfigBag) -> Option<(Cow<'static, str>, Cow<'static, str>)> {
+        let timeout_config = cfg.get::<TimeoutConfig>()?;
+        let socket_read = timeout_config.read_timeout()?;
```

Comment:
Is this right? Seems like it needs to factor in multiple possible timeouts.

Reply:
The internal-only spec for this specifies the calculation this way.